{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  T-cell vaccine design\n",
    "\n",
    "Design conserved immune-optimised vaccine(s) to induce a broad T-cell response.\n",
    "\n",
    "A graph-based method is used to design T-cell vaccines based on a set of input sequences corresponding to a (viral) target protein. A sliding window of length `k` is applied to the input sequences to split them into potential T-cell epitopes (PTEs) `e` which are represented as nodes in a graph `G`. Nodes sharing an overlap of subsequent `k - 1` amino acids are connected by an edge. The epitopes are scored based on their conservation and their likelihood of being presented by MHC molecules. The graph is then traversed to find the optimal path through the nodes, which is the set of epitopes that maximises the sum of the scores, and represents the vaccine design."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "from tvax.analyse import run_parameter_sweep\n",
    "from tvax.config import EpitopeGraphConfig\n",
    "from tvax.design import design_vaccines\n",
    "from tvax.eval import compute_population_coverage, compute_pathogen_coverage, compute_eigen_dist\n",
    "from tvax.graph import build_epitope_graph\n",
    "from tvax.plot import *\n",
    "from tvax.seq import compute_percent_match, path_to_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    'fasta_path': 'data/input/sar_rbd_protein.fasta',\n",
    "    'results_dir': 'data/results_sar_rbd',\n",
    "    'human_proteome_path': 'data/input/human_proteome_2023.03.14.fasta.gz',\n",
    "    'mhc1_alleles_path': '../optivax/scoring/MHC1_allele_mary_cleaned.txt',\n",
    "    'mhc2_alleles_path': '../optivax/scoring/MHC2_allele_marry.txt',\n",
    "    'hap_freq_mhc1_path': '../optivax/haplotype_frequency_marry.pkl',\n",
    "    'hap_freq_mhc2_path': '../optivax/haplotype_frequency_marry2.pkl',\n",
    "    'k': [9, 12],\n",
    "    'm': 1,\n",
    "    'n_target': 1,\n",
    "    'aligned': False,\n",
    "    'decycle': True,\n",
    "    'equalise_clades': True,\n",
    "    'n_clusters': None,\n",
    "    'weights': {\n",
    "        'frequency': 1,\n",
    "        'population_coverage_mhc1': 0,\n",
    "        'population_coverage_mhc2': 0\n",
    "    }\n",
    "}\n",
    "\n",
    "config = EpitopeGraphConfig(**params)\n",
    "print(config.json(indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the input epitopes\n",
    "# Cyclic example: kmers_dict = {'MSA': 0.6, 'SAM': 0.2, 'AMS': 0.4}\n",
    "kmers_dict = {\n",
    "    'MSA': {'score': 0.6},\n",
    "    'SAM': {'score': 0.2},\n",
    "    'AMQ': {'score': 0.2},\n",
    "    'MQL': {'score': 0.2},\n",
    "    'SAR': {'score': 0.4},\n",
    "    'MGA': {'score': 0.3},\n",
    "    'GAR': {'score': 0.7},\n",
    "    'ARQ': {'score': 0.4},\n",
    "    'RQL': {'score': 0.4}\n",
    "}\n",
    "\n",
    "# Construct the graph\n",
    "G = build_epitope_graph(config, kmers_dict=kmers_dict)\n",
    "\n",
    "# Find the optimal path(s) through the graph of epitopes\n",
    "Q = design_vaccines(G, config)\n",
    "print([path_to_seq(path) for path in Q])\n",
    "\n",
    "# Plot the results\n",
    "fig = plot_epitope_graph(G, Q, node_size=2000, ylim=[-0.1, 1], with_labels=True, interactive=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct the epitope graph\n",
    "Create a Directed Graph (`DiGraph`) using the `networkx` package, where each epitope `e` is a node and edges connect nodes where the last `k−1` characters of `ea` match the first `k−1` characters of `eb`. For computational convenience, two extra nodes `BEGIN` and `END` are added. The `BEGIN` node connects to all the nodes that lack predecessors (`P(e)`) (corresponding to epitopes that are the first `k` characters in a sequence). Nodes that lack successors (`S(e)`) (because they are the last `k` characters in a sequence) are connected to the `END`. For plotting convenience, the length shortest path to the `BEGIN` node is added as a node attribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "epitope_graph = build_epitope_graph(config)\n",
    "print(nx.info(epitope_graph))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design\n",
    "Take a path through the graph to optimise epitope frequency.\n",
    "\n",
    "The forward loop computes the function `F(e)` (the largest sum achievable for any path that terminates with the epitope `e`) for all the nodes in a stepwise manner. The backward loop chooses the node with maximum value as the final epitope in our optimal string and works backwards to build the path that achieves the maximal score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the optimal path(s) through the graph of epitopes\n",
    "vaccine_designs = design_vaccines(epitope_graph, config)\n",
    "[path_to_seq(path) for path in vaccine_designs]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the epitope graph\n",
    "The nodes are the epitopes `e` and the edges connect epitopes whose sequences overlap by `k − 1` amino acids. The x-axis shows the shortest path length to the `BEGIN` node, the y-axis indicates the epitope frequency `f(e)` in this target sequence set. The optimal path is shown in red which corresponds to the protein sequence that maximizes epitope coverage of the population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# F(e)\n",
    "fig = plot_epitope_graph(epitope_graph, vaccine_designs, node_size=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate the implemented scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare different scores to see if they are correlated to determine if they are redundant and if they're is a relationship between the different scores\n",
    "> Here, we can see that the frequency and immune score (MHC binding) are NOT correlated. This means that we have not identified why the epitope frequency is having a disproportionately large affect on the overall score. It also means that the immune scores are not redundant and are providing additional information that is not captured by the epitope frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plot_corr(epitope_graph)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "View the distribution of the scores for all the epitopes in the graph\n",
    "> Looking at the distribution of the epitope frequencies we can see that there are many infrequent PTEs and a few frequent PTEs. My current hypothesis is that the epitope frequency is having a disproportionately large affect on the overall score because of the **\"zero-sum\"** nature of epitope frequency. At a particular position, for one epitope to have a high frequency the other epitopes at that position must have a low frequency. Crucially, as this is for each particular position the epitope frequency has a disproportionately large affect on the overall score because it affects the choice of epitope at each decision point."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plot_score(epitope_graph, score='population_coverage_mhc2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate the vaccine design(s)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine how each of the individual scores contribute to the overall score\n",
    "> Here, we can see that the epitope frequency is having a disproportionately large affect on the overall score (despite the frequency score having a smaller weight). This is because all epitopes will have a non-zero frequency score, whereas most of the population coverage score will be zero for most epitopes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, score_dict = plot_scores(epitope_graph, config.weights, vaccine_designs, percent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for score_name, score_val in dict(config.weights).items():\n",
    "    contrib = score_dict[score_name]\n",
    "    print(f'The score {score_name} contributed to {contrib:.2f}% of the total score of the vaccine design, averaged across positions')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_target = 5\n",
    "pop_cov_mhc1 = compute_population_coverage(vaccine_designs[0], n_target, config, \"mhc1\")\n",
    "pop_cov_mhc2 = compute_population_coverage(vaccine_designs[0], n_target, config, \"mhc2\")\n",
    "path_cov = compute_pathogen_coverage(vaccine_designs[0], config)\n",
    "\n",
    "print(\n",
    "    f\"{pop_cov_mhc1 * 100:.2f}% of the population is predicted to have ≥ {n_target} Class I peptide-HLA hits produced by the vaccine\"\n",
    ")\n",
    "print(\n",
    "    f\"{pop_cov_mhc2 * 100:.2f}% of the population is predicted to have ≥ {n_target} Class II peptide-HLA hits produced by the vaccine\"\n",
    ")\n",
    "print(\n",
    "    f\"{path_cov * 100:.2f}% of potential T-cell epitopes for each target input sequence is covered by the vaccine on average\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare the generated vaccine design to a design optimised only for epitope frequency to determine if the additional scores change the vaccine design\n",
    "\n",
    "> The additional immune scores (MHC binding and immunogenicity) do not change the vaccine design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_sar_freq = 'MSDNGPQNQRSAPRITFGGPTDSTDNNQDGGRSGARPKQRRPQGLPNNTASWFTALTQHGKEELRFPRGQGVPINTNSGKDDQIGYYRRATRRVRGGDGKMKELSPRWYFYYLGTGPEASLPYGANKEGIVWVATEGALNTPKDHIGTRNPNNNAAIVLQLPQGTTLPKGFYAEGSRGGSQASSRSSSRSRGNSRNSTPGSSRGNSPARMASGGGETALALLLLDRLNQLESKVSGKGQQQQGQTVTKKSAAEASKKPRQKRTATKQYNVTQAFGRRGPEQTQGNFGDQELIRQGTDYKHWPQIAQFAPSASAFFGMSRIGMEVTPSGTWLTYHGAIKLDDKDPQFKDNVILLNKHIDAYKTFPPTEPKKDKKKKTDEAQPLPQRQKKQPTVTLLPAADMDDFSRQLQNSMSGASADSTQA'\n",
    "seq_design = path_to_seq(vaccine_designs[0])\n",
    "align, perc_match = compute_percent_match(seq_sar_freq, seq_design)\n",
    "\n",
    "print(align)\n",
    "print(f'The sequences are {perc_match:.2f}% similar')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### See where the vaccine design(s) are in sequence space\n",
    "For different numbers of clusters, plot the input target sequences and the vaccine design(s) on a PCA plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_clusters = [6]\n",
    "\n",
    "for n in n_clusters:\n",
    "    config.n_clusters = n\n",
    "    epitope_graph = build_epitope_graph(config)\n",
    "    vaccine_designs = design_vaccines(epitope_graph, config)\n",
    "    fig, comp_df = plot_vaccine_design_pca(vaccine_designs, config)\n",
    "    eigen_dist = compute_eigen_dist(comp_df)\n",
    "    print(f'Using {n} clusters, the vaccine design is {eigen_dist:.2f} standard deviations away from the mean of the input/target sequences')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Iterative clade vaccine design\n",
    "Iteratively design a vaccine by first designing a vaccine for each of the clusters and then combining the vaccine designs into a single vaccine design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tvax.graph import load_fasta, assign_clades\n",
    "\n",
    "n_clusters = [3]\n",
    "\n",
    "for n in n_clusters:\n",
    "\n",
    "    # Set the parameters\n",
    "    config.n_clusters = n\n",
    "    config.equalise_clades = True\n",
    "\n",
    "    # Load the FASTA sequences\n",
    "    seqs_dict = load_fasta(config.fasta_path)\n",
    "\n",
    "    # Assign the sequences to clusters\n",
    "    clusters_dict = assign_clades(seqs_dict, config)\n",
    "    # Organise the sequences into clusters\n",
    "    seqs_clusters_dict = {cluster: [] for cluster in clusters_dict.values()}\n",
    "    for seq, cluster in clusters_dict.items():\n",
    "        seqs_clusters_dict[cluster].append(seq)\n",
    "    \n",
    "    # Design a vaccine for each cluster\n",
    "    config.equalise_clades = False\n",
    "    vaccine_seqs_dict = {}\n",
    "    for cluster, seqs in seqs_clusters_dict.items():\n",
    "        cluster_seqs = {seq: seqs_dict[seq] for seq in seqs}\n",
    "        epitope_graph = build_epitope_graph(config, seqs_dict=cluster_seqs)\n",
    "        vaccine_designs = design_vaccines(epitope_graph, config)\n",
    "        vaccine_seqs_dict[f'cluster_{cluster}_vaccine_design'] = path_to_seq(vaccine_designs[0])\n",
    "    \n",
    "    # Generate a vaccine design using the clade designs\n",
    "    config.n_clusters = 1\n",
    "    epitope_graph = build_epitope_graph(config, seqs_dict=vaccine_seqs_dict)\n",
    "    vaccine_designs = design_vaccines(epitope_graph, config)\n",
    "\n",
    "    # Plot the PCA of where the vaccine design falls in the space of the original target sequences\n",
    "    config.n_clusters = n\n",
    "    fig = plot_vaccine_design_pca(vaccine_designs, config)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quality control - MHC binding\n",
    "Check which MHC alleles are represented in the vaccine design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alleles = None # [a.replace('*','') for a in config.alleles]\n",
    "fig, peptide_hla_hits_df = plot_mhc_heatmap(vaccine_designs, config, alleles)\n",
    "fig.show()\n",
    "peptide_hla_hits_df.style.bar(subset=['n_peptide_hla_hits'], color='#5fba7d')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform a parameter sweep\n",
    "\n",
    "Generate a set of vaccine designs for different values of `n_clusters` and `config.weights.population_coverage` and compare the designs based on the % population and pathogen coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "param_sweep_df = run_parameter_sweep(\n",
    "    n_clusters = list(range(1, 21)),\n",
    "    pop_cov_weights = list(range(1, 21)),\n",
    "    results_path = 'data/parameter_sweep_results.csv',\n",
    "    config = config\n",
    ")\n",
    "\n",
    "heatmap_fig = plot_param_sweep_heatmap(param_sweep_df)\n",
    "scatter_fig = plot_param_sweep_scatter(param_sweep_df, x=\"path_cov\", y=\"pop_cov\")\n",
    "\n",
    "heatmap_fig.show()\n",
    "scatter_fig.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "f873ee3bbd6b65e3060a1ff5e0b4755f776b262fd9907a4f86f67620d9e561eb"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
