{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  T-cell vaccine design\n",
    "Design vaccines to elicit a T-cell response by optimising coverage of potential T-cell epitope (PTEs)\n",
    "\n",
    "Here the term epitope `e` refers to a potential T-cell epitope (PTE) which is a short subsequence of k amino acids and also  represented as a node in the k-mer graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import igviz as ig\n",
    "from itertools import product\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change\n",
    "fasta_path = '../data/nucleoprotein/3_nuc_pro_uniq.fa'\n",
    "k = 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def seqs_to_kmers(seqs, k=9):\n",
    "    \"\"\"\n",
    "    Returns a dictionary of all possible k-mers and their frequencies for a given list of sequences and value of k\n",
    "    :param seqs: List of amino acid sequences\n",
    "    :param k: Integer for substring length \n",
    "    :returns: Dictionary containing all possible k-mers and their frequencies\n",
    "    \"\"\"\n",
    "    kmers_dict = {}\n",
    "    N = len(seqs)\n",
    "    for seq in seqs:\n",
    "        # Get a unique set of k-mers for each sequence with the position they were first found in\n",
    "        seq_kmers = {}\n",
    "        for i in range(N - k + 1):\n",
    "            e = seq[i:i+k]\n",
    "            if e not in seq_kmers:\n",
    "                seq_kmers[e] = i + 1\n",
    "        # Count the number of seqs for each k-mer (n)\n",
    "        for e, pos in seq_kmers.items():\n",
    "            if e in kmers_dict:\n",
    "                kmers_dict[e]['n'] = kmers_dict[e]['n'] + 1\n",
    "            else:\n",
    "                kmers_dict[e] = {'n': 1, 'pos': pos}\n",
    "    # Calculate epitope frequency\n",
    "    for e in kmers_dict:\n",
    "        kmers_dict[e]['freq'] = kmers_dict[e]['n']/N\n",
    "    return kmers_dict\n",
    "\n",
    "\n",
    "def path_to_seq(path):\n",
    "    \"\"\"\n",
    "    Returns an AA string for a list of epitopes (path)\n",
    "    \"\"\"\n",
    "    seq = [path[0]] + [e[-1] for e in path[1:]]\n",
    "    return ''.join(seq)\n",
    "\n",
    "\n",
    "def argmax(lst):\n",
    "    \"\"\"\n",
    "    Returns the index for the maximum value in a list\n",
    "    \"\"\"\n",
    "    return lst.index(max(lst))\n",
    "\n",
    "\n",
    "def P(G, e):\n",
    "    \"\"\"\n",
    "    Returns the predecessors for a given graph G and node e\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of predecessors\n",
    "    \"\"\"\n",
    "    return list(G.predecessors(e))\n",
    "\n",
    "\n",
    "def S(G, e):\n",
    "    \"\"\"\n",
    "    Returns the successors for a given graph G and node e\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of successors\n",
    "    \"\"\"\n",
    "    return list(G.successors(e))\n",
    "\n",
    "\n",
    "def f(G, e, f='Frequency'):\n",
    "    \"\"\"\n",
    "    Returns the feature for a given epitope e eg frequency in the population\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :param e: String for the node feature (default = 'Frequency')\n",
    "    :returns: Float for the epitope frequency\n",
    "    \"\"\"\n",
    "    return G.nodes[e][f]\n",
    "\n",
    "\n",
    "def F(G, e):\n",
    "    \"\"\"\n",
    "    Returns the the maximum total frequency over all paths that end in e\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: Float for the maximum total epitope frequency\n",
    "    \"\"\"\n",
    "    # Use precomputed F(e) if it already exists for the epitope\n",
    "    if 'F(e)' not in G.nodes[e]:\n",
    "        predecessors = P(G, e)\n",
    "        if not predecessors:\n",
    "            # If the set of predecessors P(e) is empty, then F(e) = f(e)\n",
    "            Fe = f(G, e)\n",
    "        else:\n",
    "            # If the set of predecessors P(e) is not empty, then F(e) = f(e) + max(F(P(e)))\n",
    "            Fe = f(G, e) + max([F(G, pe) for pe in predecessors])\n",
    "        # Save F(e) to the graph for this epitope\n",
    "        nx.set_node_attributes(G, {e: Fe}, 'F(e)')\n",
    "    return f(G, e, f='F(e)')\n",
    "\n",
    "\n",
    "def backward(G, path=[]):\n",
    "    \"\"\"\n",
    "    Returns the path that achieves the maximal score\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param path: List of epitope strings to complete (deafult=[])\n",
    "    :returns: List of epitope strings on path that achieve maximum score\n",
    "    \"\"\"\n",
    "    # Get the precomputed F(e) from the graph for all epitopes\n",
    "    Fe_dict = nx.get_node_attributes(G, 'F(e)')\n",
    "    if not path:\n",
    "        # Get the epitope with the maximum F(e) as the final epitope in our optimal path\n",
    "        path = [max(Fe_dict, key=Fe_dict.get)]\n",
    "    # Get the most recently added epitope e and it's predecessors P(e)\n",
    "    e = path[0]\n",
    "    predecessors = P(G, e)\n",
    "    if predecessors[0] != 'BEGIN':\n",
    "        # Add the best (highest F(e)) predecessor P(e) of epitope e to our path\n",
    "        i = argmax([Fe_dict[pe] for pe in predecessors])\n",
    "        path.insert(0, predecessors[i])\n",
    "        # Repeat until you get to the start\n",
    "        backward(G, path)\n",
    "    return path\n",
    "\n",
    "\n",
    "def construct_graph(kmers_dict):\n",
    "    \"\"\"\n",
    "    Return a Directed Graph with unique k-mers as nodes, where overlapping k-mers are connected by edges\n",
    "    :param kmers_dict: Dictionary containing k-mers and their counts\n",
    "    :returns: Directed Graph containing k-mers\n",
    "    \"\"\"\n",
    "    # Create graph\n",
    "    G = nx.DiGraph()\n",
    "    # Add nodes - for each unique k-mer\n",
    "    for e in kmers_dict:\n",
    "        freq = kmers_dict[e]['freq']\n",
    "        pos = kmers_dict[e]['pos']\n",
    "        G.add_node(e, Frequency=freq, pos=(pos, freq))\n",
    "    # Add edges - where the last k−1 characters of ea match the first k−1 characters of eb\n",
    "    for ea, eb in product(G.nodes(), G.nodes()):\n",
    "        if not G.has_edge(ea, eb) and ea[1:] == eb[:-1]:\n",
    "            G.add_edge(ea, eb)\n",
    "    # Add begin and end nodes\n",
    "    begin_nodes = [e for e in list(G.nodes) if not P(G, e)]\n",
    "    end_nodes = [e for e in list(G.nodes) if not S(G, e)]\n",
    "    end_pos = max([pos[0] for pos in list(nx.get_node_attributes(G, 'pos').values())]) + 1\n",
    "    G.add_node('BEGIN', Frequency=0, pos=(0, 0))\n",
    "    G.add_node('END', Frequency=0, pos=(end_pos, 0))\n",
    "    for e in begin_nodes:\n",
    "        G.add_edge('BEGIN', e)\n",
    "    for e in end_nodes:\n",
    "        G.add_edge(e, 'END')\n",
    "    return G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the input epitopes\n",
    "kmers_dict = {\n",
    "    'MSA': {'freq':6, 'pos':1},\n",
    "    'SAM': {'freq':2, 'pos':2},\n",
    "    'AMQ': {'freq':2, 'pos':3},\n",
    "    'MQL': {'freq':2, 'pos':4},\n",
    "    'SAR': {'freq':4, 'pos':2},\n",
    "    'MGA': {'freq':3, 'pos':1},\n",
    "    'GAR': {'freq':7, 'pos':2},\n",
    "    'ARQ': {'freq':4, 'pos':3},\n",
    "    'RQL': {'freq':4, 'pos':4},\n",
    "}\n",
    "\n",
    "# Construct the graph\n",
    "G = construct_graph(kmers_dict)\n",
    "\n",
    "# Forward loop - compute F(e)\n",
    "for e in G.nodes:\n",
    "    F(G, e)\n",
    "    \n",
    "# Backward loop - build the path that achieves the maximal score\n",
    "path = backward(G)\n",
    "print(path_to_seq(path))\n",
    "\n",
    "# Plot the results\n",
    "fig = ig.plot(G, color_method='Frequency', node_text=['Frequency', 'F(e)']) # , layout='spectral'\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Backward loop - build the path that achieves the maximal score\n",
    "path = backward(G)\n",
    "path_to_seq(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the FASTA sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_seqs = SeqIO.parse(open(fasta_path),'fasta')\n",
    "seqs_dict = {fasta.id: str(fasta.seq) for fasta in fasta_seqs}\n",
    "seqs = list(seqs_dict.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split into k-mers\n",
    "Compute all possible k-mers of length `k` for the given sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers_dict = seqs_to_kmers(seqs, k)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct the k-mer graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "G = construct_graph(kmers_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the k-mer graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = ig.plot(G, color_method='Frequency', node_text=['Frequency', 'pos']) # layout='spiral','spring'\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assembly \n",
    "Take a path through the graph to optimise epitope frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Add decycling\n",
    "\n",
    "# Forward loop - compute F(e)\n",
    "for e in G.nodes:\n",
    "    F(G, e)\n",
    "\n",
    "# Backward loop - build the path that achieves the maximal score\n",
    "path = backward(G)\n",
    "path_to_seq(path)\n",
    "\n",
    "# TODO: Investigate why the final sequence is so short (due to the lack of decycling?)\n",
    "# TODO: Add additional measures to scoring function eg binding affinity prediction"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
