{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  T-cell vaccine design\n",
    "Design vaccine(s) to elicit a T-cell response by optimising coverage of potential T-cell epitope (PTEs)\n",
    "\n",
    "Here the term epitope `e` refers to a potential T-cell epitope (PTE) which is a short subsequence of `k` amino acids and also  represented as a node in the epitope graph `G`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import igviz as ig\n",
    "from itertools import product\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change\n",
    "fasta_path = '../data/nucleoprotein/3_nuc_pro_uniq.fa'\n",
    "k = 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################\n",
    "# Utils to convert the format of sequences\n",
    "##########################################\n",
    "\n",
    "def seq_to_kmers(seq, k):\n",
    "    \"\"\"\n",
    "    Returns a list of k-mers of length k for a given string of amino acid sequence\n",
    "    \"\"\"\n",
    "    return [seq[i:i+k] for i in range(len(seq) - k + 1)]\n",
    "\n",
    "\n",
    "def seqs_to_kmers_dict(seqs, k=9):\n",
    "    \"\"\"\n",
    "    Returns a dictionary of all possible k-mers and their frequencies for a given list of sequences and value of k\n",
    "    :param seqs: List of amino acid sequences\n",
    "    :param k: Integer for substring length \n",
    "    :returns: Dictionary containing all possible k-mers and their frequencies\n",
    "    \"\"\"\n",
    "    kmers_dict = {}\n",
    "    N = len(seqs)\n",
    "    for seq in seqs:\n",
    "        # Get a unique set of k-mers for each sequence\n",
    "        kmers = set(seq_to_kmers(seq, k))\n",
    "        # Count the number of seqs for each k-mer (n)\n",
    "        for e in kmers:\n",
    "            if e in kmers_dict:\n",
    "                kmers_dict[e] += 1\n",
    "            else:\n",
    "                kmers_dict[e] = 1\n",
    "    # Calculate epitope frequency\n",
    "    for e, n in kmers_dict.items():\n",
    "        kmers_dict[e] = n/N\n",
    "    return kmers_dict\n",
    "\n",
    "\n",
    "def path_to_seq(path):\n",
    "    \"\"\"\n",
    "    Returns an AA string for a list of epitopes (path)\n",
    "    \"\"\"\n",
    "    seq = [path[0]] + [e[-1] for e in path[1:]]\n",
    "    return ''.join(seq)\n",
    "\n",
    "\n",
    "############################################\n",
    "# Utils to get the min/max index from a list\n",
    "############################################\n",
    "\n",
    "def argmax(lst):\n",
    "    \"\"\"\n",
    "    Returns the index for the maximum value in a list\n",
    "    \"\"\"\n",
    "    return lst.index(max(lst))\n",
    "\n",
    "\n",
    "def argmin(lst):\n",
    "    \"\"\"\n",
    "    Returns the index for the minimum value in a list\n",
    "    \"\"\"\n",
    "    return lst.index(min(lst))\n",
    "\n",
    "\n",
    "#####################################\n",
    "# Utils to retrieve info from a graph\n",
    "#####################################\n",
    "\n",
    "def P(G, e):\n",
    "    \"\"\"\n",
    "    Returns the predecessors for a given graph G and node e\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of predecessors\n",
    "    \"\"\"\n",
    "    return list(G.predecessors(e))\n",
    "\n",
    "\n",
    "def S(G, e):\n",
    "    \"\"\"\n",
    "    Returns the successors for a given graph G and node e\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of successors\n",
    "    \"\"\"\n",
    "    return list(G.successors(e))\n",
    "\n",
    "\n",
    "def f(G, e, f='Frequency'):\n",
    "    \"\"\"\n",
    "    Returns the feature for a given epitope e eg frequency in the population\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :param e: String for the node feature (default = 'Frequency')\n",
    "    :returns: Float for the epitope frequency\n",
    "    \"\"\"\n",
    "    return G.nodes[e][f]\n",
    "\n",
    "\n",
    "############################################\n",
    "# Decycling - remove all cycles from a graph\n",
    "############################################\n",
    "\n",
    "def decycle_graph(G):\n",
    "    \"\"\"\n",
    "    Return a Directed Graph with no cycles\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :returns: Directed Graph containing epitopes and no cycles\n",
    "    \"\"\"\n",
    "    # j is a list of all compnents; each component is a list of nodes in G\n",
    "    components = list(nx.strongly_connected_components(G))\n",
    "    # Discard all single node components - no cycles there!\n",
    "    components = [j for j in components if len(j) != 1]\n",
    "    if len(components) != 0:\n",
    "        for j in components:\n",
    "            # Randomly choose two nodes from the selected component\n",
    "            ea, eb = random.sample(list(j), k=2)\n",
    "            cycle = cycle_from_two_nodes(G, ea, eb)\n",
    "            if cycle:\n",
    "                ea, eb = weak_edge_in_cycle(G, cycle)\n",
    "                G.remove_edge(ea, eb)\n",
    "                # Repeat until graph is acyclic\n",
    "                G = decycle_graph(G)\n",
    "    return G\n",
    "\n",
    "\n",
    "def cycle_from_two_nodes(G, ea, eb):\n",
    "    \"\"\"\n",
    "    Returns the cycle (i.e. path that starts and ends in with the same epitope) for two nodes\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param ea: String for the first given potential T-cell epitope (PTE)\n",
    "    :param eb: String for the second given potential T-cell epitope (PTE)\n",
    "    :returns: List of epitope strings on path that is a cycle\n",
    "    \"\"\"\n",
    "    try:\n",
    "        path_ab = nx.shortest_path(G, source=ea, target=eb)\n",
    "        path_ba = nx.shortest_path(G, source=eb, target=ea)\n",
    "        # Merge two paths into a cycle\n",
    "        cycle = path_ab[:-1] + path_ba\n",
    "    except nx.NetworkXNoPath:\n",
    "        cycle = []\n",
    "    return cycle\n",
    "\n",
    "\n",
    "def weak_edge_in_cycle(G, cycle):\n",
    "    \"\"\"\n",
    "    Returns the weak edge (edge with the lowest score) in a cycle\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param cycle: List of epitope strings on path that is a cycle\n",
    "    :returns: Tuple for the weak edge containing the two epitope strings\n",
    "    \"\"\"\n",
    "    edges = seq_to_kmers(cycle, k=2)\n",
    "    values = []\n",
    "    for ea, eb in edges:\n",
    "        # v is heuristic “value” of edge\n",
    "        v = f(G, ea) + f(G, eb)\n",
    "        # Add value if cutting edge would isolate ea\n",
    "        if len(S(G, ea)) == 1:\n",
    "            v = v + f(G, ea)\n",
    "        # Add value if cutting edge would isolate eb\n",
    "        if len(P(G, eb)) == 1:\n",
    "            v = v + f(G, eb)\n",
    "        values.append(v)\n",
    "    ea, eb = edges[argmin(values)]\n",
    "    return ea, eb\n",
    "\n",
    "\n",
    "###############################################\n",
    "# Find optimal path through a graph of epitopes\n",
    "###############################################\n",
    "\n",
    "def find_optimal_path(G):\n",
    "    \"\"\"\n",
    "    Returns the optimal path through a graph of epitopes\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :returns: List of epitope strings on the optimal path\n",
    "    \"\"\"\n",
    "    # Forward loop - compute F(e)\n",
    "    for e in G.nodes:\n",
    "        F(G, e)\n",
    "    # Backward loop - build the path that achieves the maximal score\n",
    "    path = backward(G)\n",
    "    return path\n",
    "\n",
    "\n",
    "def F(G, e):\n",
    "    \"\"\"\n",
    "    Returns the maximum total frequency over all paths that end in e\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: Float for the maximum total epitope frequency\n",
    "    \"\"\"\n",
    "    # Use precomputed F(e) if it already exists for the epitope\n",
    "    if 'F(e)' not in G.nodes[e]:\n",
    "        predecessors = P(G, e)\n",
    "        if not predecessors:\n",
    "            # If the set of predecessors P(e) is empty, then F(e) = f(e)\n",
    "            Fe = f(G, e)\n",
    "        else:\n",
    "            # If the set of predecessors P(e) is not empty, then F(e) = f(e) + max(F(P(e)))\n",
    "            Fe = f(G, e) + max([F(G, pe) for pe in predecessors])\n",
    "        # Save F(e) to the graph for this epitope\n",
    "        nx.set_node_attributes(G, {e: Fe}, 'F(e)')\n",
    "    return f(G, e, f='F(e)')\n",
    "\n",
    "\n",
    "def backward(G, path=[]):\n",
    "    \"\"\"\n",
    "    Returns the path that achieves the maximal score\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param path: List of epitope strings to complete (deafult=[])\n",
    "    :returns: List of epitope strings on path that achieve maximum score\n",
    "    \"\"\"\n",
    "    # Get the precomputed F(e) from the graph for all epitopes\n",
    "    Fe_dict = nx.get_node_attributes(G, 'F(e)')\n",
    "    if not path:\n",
    "        # Get the epitope with the maximum F(e) as the final epitope in our optimal path\n",
    "        path = [max(Fe_dict, key=Fe_dict.get)]\n",
    "    # Get the most recently added epitope e and it's predecessors P(e)\n",
    "    e = path[0]\n",
    "    predecessors = P(G, e)\n",
    "    if predecessors[0] != 'BEGIN':\n",
    "        # Add the best (highest F(e)) predecessor P(e) of epitope e to our path\n",
    "        i = argmax([Fe_dict[pe] for pe in predecessors])\n",
    "        path.insert(0, predecessors[i])\n",
    "        # Repeat until you get to the start\n",
    "        backward(G, path)\n",
    "    return path\n",
    "\n",
    "\n",
    "###########################################################\n",
    "# Cocktail: Find (and iteratively refine) a set of antigens\n",
    "###########################################################\n",
    "\n",
    "def cocktail(G, m):\n",
    "    \"\"\"\n",
    "    Returns a list of m antigens\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param m: Integer for number of antigens\n",
    "    :returns: List containing m antigens\n",
    "    \"\"\"\n",
    "    Q = [] # vaccine\n",
    "    for n in range(0, m):\n",
    "        # Compute and save next antigen sequence\n",
    "        q = find_optimal_path(G)\n",
    "        # Add q to vaccine\n",
    "        Q.append(q)\n",
    "        # Save original epitope frequency for iterative refinement\n",
    "        freq_dict = nx.get_node_attributes(G, 'Frequency')\n",
    "        nx.set_node_attributes(G, freq_dict, 'Original Frequency')\n",
    "        # No credit for including e in subsequent antigens\n",
    "        for e in q:\n",
    "            nx.set_node_attributes(G, {e: 0}, 'Frequency')\n",
    "        # Remove F(e) so it's recomputed using the updated frequencies\n",
    "        for (n,d) in G.nodes(data=True):\n",
    "            del d['F(e)']\n",
    "    # TODO: Optional - Repeat for iterative refinement\n",
    "    return Q\n",
    "\n",
    "\n",
    "#######################################\n",
    "# Construct/visualise the epitope graph\n",
    "#######################################\n",
    "\n",
    "def construct_graph(kmers_dict, decycle=True, edge_colour='#BFBFBF'):\n",
    "    \"\"\"\n",
    "    Return a Directed Graph with unique k-mers as nodes, where overlapping k-mers are connected by edges\n",
    "    :param kmers_dict: Dictionary containing k-mers and their counts\n",
    "    :param decycle: Boolean if the output graph should be decycled (default=True)\n",
    "    :param edge_colour: String for edge colour (default='black')\n",
    "    :returns: Directed Graph containing epitopes\n",
    "    \"\"\"\n",
    "    # Create graph\n",
    "    G = nx.DiGraph()\n",
    "    # Add nodes - for each unique k-mer\n",
    "    for e, freq in kmers_dict.items():\n",
    "        G.add_node(e, Frequency=freq)\n",
    "    # Add edges - where the last k−1 characters of ea match the first k−1 characters of eb\n",
    "    for ea, eb in product(G.nodes(), G.nodes()):\n",
    "        if not G.has_edge(ea, eb) and ea[1:] == eb[:-1]:\n",
    "            G.add_edge(ea, eb, colour=edge_colour)\n",
    "    # Decycle graph\n",
    "    if decycle:\n",
    "        G = decycle_graph(G)\n",
    "    # Add begin and end nodes\n",
    "    begin_nodes = [e for e in list(G.nodes) if not P(G, e)]\n",
    "    end_nodes = [e for e in list(G.nodes) if not S(G, e)]\n",
    "    G.add_node('BEGIN', Frequency=0)\n",
    "    G.add_node('END', Frequency=0)\n",
    "    for e in begin_nodes:\n",
    "        G.add_edge('BEGIN', e, colour=edge_colour)\n",
    "    for e in end_nodes:\n",
    "        G.add_edge(e, 'END', colour=edge_colour)\n",
    "    # Add the position node attribute\n",
    "    for e in G.nodes():\n",
    "        freq = f(G, e)\n",
    "        pos = nx.shortest_path_length(G, source='BEGIN', target=e)\n",
    "        nx.set_node_attributes(G, {e:(pos, freq)}, 'pos')\n",
    "    # TODO: Make sure END node is at the end\n",
    "    # end_pos = max([pos[0] for pos in list(nx.get_node_attributes(G, 'pos').values())]) + 1\n",
    "    return G\n",
    "\n",
    "\n",
    "def plot_graph(G, path=None, node_size=150, with_labels=False, ylim=([0,1]), interactive=False):\n",
    "    \"\"\"\n",
    "    Plot the epitope graph\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param path: List of epitope strings (deafult=None)\n",
    "    :param node_size: Integer for size of nodes in non-interactive plot (default=150)\n",
    "    :param with_labels: Boolean for if epitope labels should be displayed in the non-interactive plot (default=False)\n",
    "    :param ylim: List for y-axis limits (default=[0,1])\n",
    "    :param interactive: Boolean for if the plot should be interactive (default=False)\n",
    "    :returns: None\n",
    "    \"\"\"\n",
    "    if interactive:\n",
    "        fig = ig.plot(G, color_method='Frequency', node_text=['Frequency', 'F(e)']) # layout='spectral','spiral','spring'\n",
    "        return fig.show()\n",
    "    else:\n",
    "        # Define vars\n",
    "        freq = list(nx.get_node_attributes(G,'Frequency').values())\n",
    "        pos = nx.get_node_attributes(G, 'pos')\n",
    "        if path:\n",
    "            path = ['BEGIN'] + path + ['END']\n",
    "            for i in range(0,len(path)-1):\n",
    "                G.edges[path[i], path[i+1]]['colour'] = 'red'\n",
    "        edge_colours = nx.get_edge_attributes(G, 'colour')\n",
    "        # Plot\n",
    "        fig, ax = plt.subplots(1,figsize=(16,8))\n",
    "        nx.draw(G, node_color=freq, pos=pos, node_size=node_size, with_labels=with_labels, edge_color=edge_colours.values(), width=2, font_color='white', ax=ax)\n",
    "        limits=plt.axis('on')\n",
    "        max_pos = max([p[0] for p in pos.values()]) + 0.5\n",
    "        ax.set_xlim([-0.5, max_pos])\n",
    "        ax.set_ylim(ylim)\n",
    "        ax.spines.right.set_visible(False)\n",
    "        ax.spines.top.set_visible(False)\n",
    "        ax.tick_params(left=True, bottom=True, labelleft=True, labelbottom=True)\n",
    "        ax.tick_params(axis='both', which='major', labelsize=14)\n",
    "        plt.ylabel('Epitope Frequency, f(e)', fontsize=18)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the input epitopes\n",
    "# Cyclic example: kmers_dict = {'MSA': 0.6, 'SAM': 0.2, 'AMS': 0.4}\n",
    "kmers_dict = {\n",
    "    'MSA': 0.6,\n",
    "    'SAM': 0.2,\n",
    "    'AMQ': 0.2,\n",
    "    'MQL': 0.2,\n",
    "    'SAR': 0.4,\n",
    "    'MGA': 0.3,\n",
    "    'GAR': 0.7,\n",
    "    'ARQ': 0.4,\n",
    "    'RQL': 0.4,\n",
    "}\n",
    "\n",
    "# Construct the graph\n",
    "G = construct_graph(kmers_dict)\n",
    "\n",
    "# Find the optimal path through the graph of epitopes\n",
    "path = find_optimal_path(G)\n",
    "print(path_to_seq(path))\n",
    "\n",
    "# Plot the results\n",
    "plot_graph(G, path=path, node_size=2000, ylim=[-0.1, 1], with_labels=True, interactive=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the FASTA sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_seqs = SeqIO.parse(open(fasta_path),'fasta')\n",
    "seqs = [str(seq.seq) for seq in fasta_seqs]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split into k-mers\n",
    "Compute all possible k-mers of length `k` for the given target sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers_dict = seqs_to_kmers_dict(seqs, k)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct the epitope graph\n",
    "Create a Directed Graph (`DiGraph`) using the `networkx` package, where each epitope `e` is a node and edges connect nodes where the last `k−1` characters of `ea` match the first `k−1` characters of `eb`. For computational convenience, two extra nodes `BEGIN` and `END` are added. The `BEGIN` node connects to all the nodes that lack predecessors (`P(e)`) (corresponding to epitopes that are the first `k` characters in a sequence). Nodes that lack successors (`S(e)`) (because they are the last `k` characters in a sequence) are connected to the `END`. For plotting convenience, the length shortest path to the `BEGIN` node is added as a node attribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "G = construct_graph(kmers_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assembly \n",
    "Take a path through the graph to optimise epitope frequency.\n",
    "\n",
    "The forward loop computes the function `F(e)` (the largest sum achievable for any path that terminates with the epitope `e`) for all the nodes in a stepwise manner. The backward loop chooses the node with maximum value as the final epitope in our optimal string and works backwards to build the path that achieves the maximal score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Add additional measures to scoring function eg binding affinity prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the optimal path through the graph of epitopes\n",
    "path = find_optimal_path(G)\n",
    "path_to_seq(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the epitope graph\n",
    "The nodes are the epitopes `e` and the edges connect epitopes whose sequences overlap by `k − 1` amino acids. The x-axis shows the shortest path length to the `BEGIN` node, the y-axis indicates the epitope frequency `f(e)` in this target sequence set. The optimal path is shown in red which corresponds to the protein sequence that maximizes epitope coverage of the population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_graph(G, path=path, node_size=25, with_labels=False, ylim=[0, 1], interactive=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
