{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  T-cell vaccine design\n",
    "Design vaccines to elicit a T-cell response by optimising coverage of potential T-cell epitope (PTEs)\n",
    "\n",
    "In this notebook the terms k-mers, epitopes, node are used interchangeably because each k-mer is a potential T-cell epitope (PTE) and is represented as a node in the k-mer graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import igviz as ig\n",
    "from itertools import product\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change\n",
    "fasta_path = '../data/nucleoprotein/3_nuc_pro_uniq.fa'\n",
    "k = 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def seqs_to_kmers(seqs, k=9):\n",
    "    \"\"\"\n",
    "    Returns a dictionary of all possible k-mers and their frequencies for a given list of sequences and value of k\n",
    "    :param seqs: List of amino acid sequences\n",
    "    :param k: Integer for substring length \n",
    "    :returns: Dictionary containing all possible k-mers and their frequencies\n",
    "    \"\"\"\n",
    "    kmers_dict = {}\n",
    "    N = len(seqs)\n",
    "    for seq in seqs:\n",
    "        # Get a unique set of k-mers for each sequence\n",
    "        kmers = set([seq[i:i+k] for i in range(N - k + 1)])\n",
    "        # Count the number of seqs for each k-mer (n)\n",
    "        for kmer in kmers:\n",
    "            if kmer in kmers_dict:\n",
    "                kmers_dict[kmer] += 1\n",
    "            else:\n",
    "                kmers_dict[kmer] = 1\n",
    "    # Calculate epitope frequency\n",
    "    for kmer, n in kmers_dict.items():\n",
    "        kmers_dict[kmer] = n/N\n",
    "    return kmers_dict\n",
    "\n",
    "\n",
    "def P(G, e):\n",
    "    \"\"\"\n",
    "    Returns the predecessors for a given graph G and node e\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of predecessors\n",
    "    \"\"\"\n",
    "    return list(G.predecessors(e))\n",
    "\n",
    "\n",
    "def S(G, e):\n",
    "    \"\"\"\n",
    "    Returns the successors for a given graph G and node e\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of successors\n",
    "    \"\"\"\n",
    "    return list(G.successors(e))\n",
    "\n",
    "\n",
    "def f(G, e, f='Frequency'):\n",
    "    \"\"\"\n",
    "    Returns the frequency for a given epitope e in the population\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :param e: String for node feature (default = 'Frequency')\n",
    "    :returns: Float for the epitope frequency\n",
    "    \"\"\"\n",
    "    return G.nodes[e][f]\n",
    "\n",
    "\n",
    "def F(G, e):\n",
    "    \"\"\"\n",
    "    Returns the the maximum total frequency over all paths that end in e\n",
    "    :param G: Directed Graph containing k-mers\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: Float for the maximum total epitope frequency\n",
    "    \"\"\"\n",
    "    if 'F(e)' not in G.nodes[e]:\n",
    "        predecessors = P(G, e)\n",
    "        if not predecessors:\n",
    "            Fe = f(G, e)\n",
    "        else:\n",
    "            Fe = f(G, e) + max([F(G, pe) for pe in predecessors])\n",
    "        nx.set_node_attributes(G, {e: Fe}, 'F(e)')\n",
    "    return f(G, e, f='F(e)')\n",
    "\n",
    "\n",
    "def construct_graph(kmers_dict):\n",
    "    \"\"\"\n",
    "    Return a Directed Graph with unique k-mers as nodes, where overlapping k-mers are connected by edges\n",
    "    :param kmers_dict: Dictionary containing k-mers and their counts\n",
    "    :returns: Directed Graph containing k-mers\n",
    "    \"\"\"\n",
    "    # Create graph\n",
    "    G = nx.DiGraph()\n",
    "    # Add nodes - for each unique k-mer\n",
    "    for kmer, count in kmers_dict.items():\n",
    "        G.add_node(kmer, Frequency=count)\n",
    "    # Add edges - where the last k−1 characters ofea match the first k−1 characters of eb\n",
    "    for n1, n2 in product(G.nodes(), G.nodes()):\n",
    "        if not G.has_edge(n1, n2) and n1[1:] == n2[:-1]:\n",
    "            G.add_edge(n1, n2)\n",
    "    # Add begin and end nodes\n",
    "    begin_nodes = [e for e in list(G.nodes) if not P(G, e)]\n",
    "    end_nodes = [e for e in list(G.nodes) if not S(G, e)]\n",
    "    G.add_node('BEGIN', Frequency=0)\n",
    "    G.add_node('END', Frequency=0)\n",
    "    for n in begin_nodes:\n",
    "        G.add_edge('BEGIN', n)\n",
    "    for n in end_nodes:\n",
    "        G.add_edge(n, 'END')\n",
    "    return G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers_dict = {\n",
    "    'MSA': 6,\n",
    "    'SAM': 2,\n",
    "    'AMQ': 2,\n",
    "    'MQL': 2,\n",
    "    'VGA': 4,\n",
    "    'SAR': 4,\n",
    "    'ARG': 7,\n",
    "    'RGL': 7,\n",
    "    'MGA': 3,\n",
    "    'GAR': 7,\n",
    "    'ARQ': 4,\n",
    "    'RQL': 4\n",
    "}\n",
    "\n",
    "G = construct_graph(kmers_dict)\n",
    "fig = ig.plot(G, color_method='Frequency', node_text=['Frequency'], layout='spectral')\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the FASTA sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_seqs = SeqIO.parse(open(fasta_path),'fasta')\n",
    "seqs_dict = {fasta.id: str(fasta.seq) for fasta in fasta_seqs}\n",
    "seqs = list(seqs_dict.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split into k-mers\n",
    "Compute all possible k-mers of length `k` for the given sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers_dict = seqs_to_kmers(seqs, k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for kmer, freq in kmers_dict.items():\n",
    "    print(kmer + \"\\t\" + str(freq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct the k-mer graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "G = construct_graph(kmers_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the k-mer graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = ig.plot(G, color_method='Frequency', node_text=['Frequency'], layout='spring') # 'spiral'\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import matplotlib.pyplot as plt\n",
    "# freq = list(nx.get_node_attributes(G,'frequency').values())\n",
    "# nx.draw(G, node_color=freq, node_size=10)\n",
    "# plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
