{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  T-cell vaccine design\n",
    "Design vaccine(s) to elicit a T-cell response by optimising coverage of potential T-cell epitope (PTEs)\n",
    "\n",
    "Here the term epitope `e` refers to a potential T-cell epitope (PTE) which is a short subsequence of `k` amino acids and also  represented as a node in the epitope graph `G`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from bin.predict_immunogenicity import Prediction # TODO: Add this to the repo\n",
    "\n",
    "import igviz as ig\n",
    "from itertools import product\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import mhcflurry\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random\n",
    "from sklearn import preprocessing\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change\n",
    "fasta_path = '../data/test/1_nuc_cds_uniq.fa'\n",
    "hla_path = '../data/test/hla_ref_set.class_i.txt'\n",
    "k = 9\n",
    "m = 1 # number of antigens\n",
    "aligned = False\n",
    "decycle = True\n",
    "weights = {\n",
    "    # 'evo_rate': 0,\n",
    "    'freq': 1,\n",
    "    'immunogenicity': 1,\n",
    "    'mhc_binding': 1,\n",
    "}\n",
    "\n",
    "# pre_msa_path = 'hyphy/pre-msa.bf'\n",
    "# post_msa_path = 'hyphy/post-msa.bf'\n",
    "# fasta_nt_path = fasta_path + '_nuc.fas'\n",
    "# fasta_pr_path = fasta_path + '_protein.fas'\n",
    "# msa_pr_path = fasta_path + '_protein.msa'\n",
    "# msa_nt_path = fasta_path + '_nuc.msa'\n",
    "# tree_path = fasta_path + '.treefile'\n",
    "# fel_path = msa_nt_path + '.FEL.json'\n",
    "# slac_path = msa_nt_path + '.SLAC.json'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "supported_alleles = !mhcflurry-predict --list-supported-alleles\n",
    "supported_alleles = list(supported_alleles)\n",
    "\n",
    "alleles = []\n",
    "with open(hla_path, 'r') as f:\n",
    "    for line in f:\n",
    "        allele = line.strip().split(',')[0]\n",
    "        if allele in supported_alleles:\n",
    "            alleles.append(allele)\n",
    "        else:\n",
    "            print(f'Allele {allele} not supported by mhcflurry')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "# Calculate scores\n",
    "##################\n",
    "\n",
    "def calc_score(kmers_dict, weights, kmer):\n",
    "    \"\"\"\n",
    "    Returns the overall score for a given k-mer\n",
    "    :param kmers_dict: Dictionary containing all possible k-mers with their corresponding scores\n",
    "    :param weights: Dictionary containing keys for each scoring criteria and values corresponding to their weight\n",
    "    :param kmer: String of k-mer\n",
    "    :returns: Float of overall score for a given k-mer\n",
    "    \"\"\"\n",
    "    return sum([weights[weight] * kmers_dict[kmer][weight] for weight in weights]) / sum(list(weights.values()))\n",
    "\n",
    "\n",
    "def predict_immunogenicity(peptides):\n",
    "    \"\"\"\n",
    "    Returns the immunogenicity score for a given list of peptides\n",
    "    :param peptides: List of peptides\n",
    "    :returns: Dictionary of peptide to immunogenicity score\n",
    "    \"\"\"\n",
    "    # TODO: Add alleles as a parameter\n",
    "    predictor = Prediction()\n",
    "    immunogenicity = predictor.main(peptides=peptides)\n",
    "    df = pd.DataFrame(immunogenicity[1:], columns=immunogenicity[0])\n",
    "    scaler = preprocessing.MinMaxScaler()\n",
    "    df['score'] = scaler.fit_transform(df['score'].values.reshape(-1, 1))\n",
    "    return df[['peptide', 'score']].set_index('peptide').to_dict()['score']\n",
    "\n",
    "\n",
    "def predict_mhc_binding(peptides, alleles, aligned=False, percentile_threshold=5):\n",
    "    \"\"\"\n",
    "    Returns the MHC binding score for a given list of peptides and alleles\n",
    "    :param peptides: List of peptides\n",
    "    :param alleles: List of alleles\n",
    "    :param aligned: Boolean indicating if the peptides are aligned\n",
    "    :param percentile_threshold: Integer indicating the percentile threshold for MHC binding\n",
    "    :returns: Dictionary of peptide to MHC binding score\n",
    "    \"\"\"\n",
    "    !mhcflurry-downloads fetch models_class1_presentation\n",
    "    predictor = mhcflurry.Class1PresentationPredictor.load()\n",
    "    if aligned:\n",
    "        peptides_dict = {e.replace('-',''):e for e in peptides}\n",
    "        peptides = list(peptides_dict.keys())\n",
    "    preds = [predictor.predict(peptides, [allele], include_affinity_percentile=True, verbose=0) for allele in alleles]\n",
    "    preds_df = pd.concat(preds, axis=0)\n",
    "    if aligned:\n",
    "        preds_df['peptide'] = preds_df['peptide'].map(peptides_dict)\n",
    "    preds_df['score'] = preds_df['affinity_percentile'].apply(lambda x: 1 if x < percentile_threshold else 0)\n",
    "    preds_df = preds_df.groupby('peptide').sum().reset_index() #.median()\n",
    "    preds_df['score'] = preds_df['score'] / len(alleles)\n",
    "    # preds_df['score'] = 1 - preds_df['affinity_percentile'] / 100\n",
    "    return preds_df[['peptide', 'score']].set_index('peptide').to_dict()['score']\n",
    "\n",
    "# def get_lrt(fel_path):\n",
    "#     \"\"\"\n",
    "#     Returns the likelihood-ratio test (LRT) with sign (+/-) to indicate if the codon is under posotive or negative selection for each codon\n",
    "#     :param fel_path: String path to HyPhy FEL JSON ouput\n",
    "#     :returns: List of LRT values for each codon with sign to indicate selection type\n",
    "#     \"\"\"\n",
    "#     fel = json.load(open(fel_path))\n",
    "#     headers = [header[0] for header in fel['MLE']['headers']] # fel['MLE']['headers']\n",
    "#     content = fel['MLE']['content']['0']\n",
    "#     df = pd.DataFrame(content, columns=headers)\n",
    "#     df['omega'] = df['beta'] / df['alpha']\n",
    "#     df['LRT_selection'] = preprocessing.normalize([np.array(df['LRT'])])[0]\n",
    "#     df['LRT_selection'] = df.apply(lambda x: x['LRT_selection'] if x['alpha'] < x['beta'] else x['LRT_selection']*-1, axis=1)\n",
    "#     return df['LRT_selection'].tolist()\n",
    "\n",
    "\n",
    "def get_evo_rate(slac_path):\n",
    "    \"\"\"\n",
    "    Returns the evolution rate score which is the the absolute number of synonymous (-1) + nonsynonymous (1) sites at each position in the alignment\n",
    "    :param slac_path: String path to HyPhy SLAC JSON ouput\n",
    "    :returns: List of evolution rate score for each codon\n",
    "    \"\"\"\n",
    "    slac = json.load(open(slac_path))\n",
    "    headers = [header[0] for header in slac['MLE']['headers']]\n",
    "    content = slac['MLE']['content']['0']['by-site']['RESOLVED'] #['AVERAGED']\n",
    "    df = pd.DataFrame(content, columns=headers)[['S', 'N']]\n",
    "    # This is a bit counter-intuitive but we set the number of non-synonymous sites (N) to negative even though they're under positive selection\n",
    "    # This is because we want to select for sites that are under negative (aka purifying) selection\n",
    "    df['N'] = df['N']*-1\n",
    "    df['evo_rate'] = df['S'] + df['N']\n",
    "    scaler = preprocessing.MinMaxScaler()\n",
    "    df['evo_rate'] = scaler.fit_transform(df['evo_rate'].values.reshape(-1, 1))\n",
    "    return df['evo_rate'].tolist()\n",
    "\n",
    "\n",
    "##########################################\n",
    "# Utils to convert the format of sequences\n",
    "##########################################\n",
    "\n",
    "def seq_to_kmers(seq, k):\n",
    "    \"\"\"\n",
    "    Returns a list of k-mers of length k for a given string of amino acid sequence\n",
    "    \"\"\"\n",
    "    return [seq[i:i+k] for i in range(len(seq) - k + 1)]\n",
    "\n",
    "\n",
    "# For aligned sequences\n",
    "def seqs_dict_to_df(seqs_dict, evo_rate):\n",
    "    \"\"\"\n",
    "    Returns a dataframe of sequences, their names and corresponding evolutionary rate for each codon\n",
    "    :param seqs_dict: Dictionary where the keys are sequence IDs and the values are aligned amino acid sequence strings\n",
    "    :param evo_rate: List containing the evolutionary rate for each codon/site in the MSA \n",
    "    :returns: Dataframe of sequences, their names and corresponding evolutionary rate for each codon\n",
    "    \"\"\"\n",
    "    seqs = [list(seq) for seq in list(seqs_dict.values())]\n",
    "    seqs.insert(0, evo_rate)\n",
    "    names = ['Evo_Rate'] + list(seqs_dict.keys())\n",
    "    df = pd.DataFrame(seqs)\n",
    "    df.index = names\n",
    "    return df\n",
    "\n",
    "\n",
    "def seqs_df_to_kmers_dict(df, weights, k=9):\n",
    "    \"\"\"\n",
    "    Returns a dictionary of all possible k-mers and their evolutionary rate, position, frequency and overall score\n",
    "    :param df: Dataframe of amino acid sequences, their names and corresponding evolutionary rate for each codon\n",
    "    :param k: Integer for substring length\n",
    "    :param weights: Dictionary containing keys for each scoring criteria and values corresponding to their weight\n",
    "    :returns: Dictionary containing all possible k-mers with their evolutionary rate, position, frequency and overall score\n",
    "    \"\"\"\n",
    "    kmers_dict = {}\n",
    "    for i in range(len(df.columns) - k + 1):\n",
    "        kmers = df.loc[:, i:i+k-1]\n",
    "        evo_rate = kmers.loc['Evo_Rate'].tolist()\n",
    "        evo_rate = sum(evo_rate) / len(evo_rate)\n",
    "        pos = i+1\n",
    "        for index, row in kmers.iterrows():\n",
    "            if index != 'Evo_Rate':\n",
    "                kmer = ''.join(row.tolist())\n",
    "                if kmer not in kmers_dict:\n",
    "                    kmers_dict[kmer] = {'evo_rate':evo_rate, 'pos':pos, 'freq':1}\n",
    "                else:\n",
    "                    if pos == kmers_dict[kmer]['pos']:\n",
    "                        kmers_dict[kmer]['freq'] += 1\n",
    "                    else:\n",
    "                        warnings.warn(f'k-mer {kmer} found at multiple positions {pos},{kmers_dict[kmer][\"pos\"]}. This was not expected to happen and may cause issues.')\n",
    "    for kmer in kmers_dict:\n",
    "        # Convert sequence counts to frequency\n",
    "        kmers_dict[kmer]['freq'] = kmers_dict[kmer]['freq'] / len(df.columns)\n",
    "        # Calculate score using weights\n",
    "        kmers_dict[kmer]['score'] = calc_score(kmers_dict, weights, kmer)\n",
    "    return kmers_dict\n",
    "\n",
    "\n",
    "# For unaligned sequences\n",
    "def seqs_to_kmers_dict(seqs, k=9, weights=weights, alleles=None, aligned=False):\n",
    "    \"\"\"\n",
    "    Returns a dictionary of all possible k-mers and their frequencies for a given list of sequences and value of k\n",
    "    :param seqs: List of amino acid sequences\n",
    "    :param k: Integer for substring length \n",
    "    :param alleles: List of MHC alleles to use for MHC binding prediction\n",
    "    :param aligned: Boolean indicating whether the sequences are aligned or unaligned\n",
    "    :returns: Dictionary containing all possible k-mers and their frequencies\n",
    "    \"\"\"\n",
    "    kmers_dict = {}\n",
    "    N = len(seqs)\n",
    "    for seq in seqs:\n",
    "        # Get a unique set of k-mers for each sequence\n",
    "        kmers = set(seq_to_kmers(seq, k))\n",
    "        # Count the number of seqs for each k-mer (n)\n",
    "        for e in kmers:\n",
    "            if e in kmers_dict:\n",
    "                kmers_dict[e]['freq'] += 1\n",
    "            else:\n",
    "                kmers_dict[e] = {'freq': 1, 'pos': 0, 'evo_rate': 0, 'immunogenicity':0, 'mhc_binding': 0, 'score': 0}\n",
    "\n",
    "    immunogenicity_dict = {e:0 for e in kmers_dict}\n",
    "    mhc_binding_dict = {e:0 for e in kmers_dict}\n",
    "    peptides = list(kmers_dict.keys())\n",
    "\n",
    "    if weights['immunogenicity']:\n",
    "        immunogenicity_dict = predict_immunogenicity(peptides)\n",
    "        \n",
    "    if weights['mhc_binding']:\n",
    "        mhc_binding_dict = predict_mhc_binding(peptides, alleles, aligned)\n",
    "\n",
    "    for e, d in kmers_dict.items():\n",
    "        kmers_dict[e]['freq'] = d['freq']/N\n",
    "        kmers_dict[e]['immunogenicity'] = immunogenicity_dict[e]\n",
    "        kmers_dict[e]['mhc_binding'] = mhc_binding_dict[e]\n",
    "        kmers_dict[e]['score'] = calc_score(kmers_dict, weights, e)\n",
    "\n",
    "    return kmers_dict\n",
    "\n",
    "\n",
    "def path_to_seq(path):\n",
    "    \"\"\"\n",
    "    Returns an AA string for a list of epitopes (path)\n",
    "    \"\"\"\n",
    "    seq = [path[0]] + [e[-1] for e in path[1:]]\n",
    "    return ''.join(seq)\n",
    "\n",
    "\n",
    "############################################\n",
    "# Utils to get the min/max index from a list\n",
    "############################################\n",
    "\n",
    "def argmax(lst):\n",
    "    \"\"\"\n",
    "    Returns the index for the maximum value in a list\n",
    "    \"\"\"\n",
    "    return lst.index(max(lst))\n",
    "\n",
    "\n",
    "def argmin(lst):\n",
    "    \"\"\"\n",
    "    Returns the index for the minimum value in a list\n",
    "    \"\"\"\n",
    "    return lst.index(min(lst))\n",
    "\n",
    "\n",
    "#####################################\n",
    "# Utils to retrieve info from a graph\n",
    "#####################################\n",
    "\n",
    "def P(G, e):\n",
    "    \"\"\"\n",
    "    Returns the predecessors for a given graph G and node e\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of predecessors\n",
    "    \"\"\"\n",
    "    return list(G.predecessors(e))\n",
    "\n",
    "\n",
    "def S(G, e):\n",
    "    \"\"\"\n",
    "    Returns the successors for a given graph G and node e\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: List of successors\n",
    "    \"\"\"\n",
    "    return list(G.successors(e))\n",
    "\n",
    "\n",
    "def f(G, e, f='Score'):\n",
    "    \"\"\"\n",
    "    Returns the feature for a given epitope e eg frequency or score in the population\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :param e: String for the node feature (default = 'Score')\n",
    "    :returns: Float for the epitope feature eg score\n",
    "    \"\"\"\n",
    "    return G.nodes[e][f]\n",
    "\n",
    "\n",
    "############################################\n",
    "# Decycling - remove all cycles from a graph\n",
    "############################################\n",
    "\n",
    "def decycle_graph(G):\n",
    "    \"\"\"\n",
    "    Return a Directed Graph with no cycles\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :returns: Directed Graph containing epitopes and no cycles\n",
    "    \"\"\"\n",
    "    # j is a list of all compnents; each component is a list of nodes in G\n",
    "    components = list(nx.strongly_connected_components(G))\n",
    "    # Discard all single node components - no cycles there!\n",
    "    components = [j for j in components if len(j) != 1]\n",
    "    if len(components) != 0:\n",
    "        for j in components:\n",
    "            # Randomly choose two nodes from the selected component\n",
    "            ea, eb = random.sample(list(j), k=2)\n",
    "            cycle = cycle_from_two_nodes(G, ea, eb)\n",
    "            if cycle:\n",
    "                ea, eb = weak_edge_in_cycle(G, cycle)\n",
    "                G.remove_edge(ea, eb)\n",
    "                # Repeat until graph is acyclic\n",
    "                G = decycle_graph(G)\n",
    "    return G\n",
    "\n",
    "\n",
    "def cycle_from_two_nodes(G, ea, eb):\n",
    "    \"\"\"\n",
    "    Returns the cycle (i.e. path that starts and ends in with the same epitope) for two nodes\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param ea: String for the first given potential T-cell epitope (PTE)\n",
    "    :param eb: String for the second given potential T-cell epitope (PTE)\n",
    "    :returns: List of epitope strings on path that is a cycle\n",
    "    \"\"\"\n",
    "    try:\n",
    "        path_ab = nx.shortest_path(G, source=ea, target=eb)\n",
    "        path_ba = nx.shortest_path(G, source=eb, target=ea)\n",
    "        # Merge two paths into a cycle\n",
    "        cycle = path_ab[:-1] + path_ba\n",
    "    except nx.NetworkXNoPath:\n",
    "        cycle = []\n",
    "    return cycle\n",
    "\n",
    "\n",
    "def weak_edge_in_cycle(G, cycle):\n",
    "    \"\"\"\n",
    "    Returns the weak edge (edge with the lowest score) in a cycle\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param cycle: List of epitope strings on path that is a cycle\n",
    "    :returns: Tuple for the weak edge containing the two epitope strings\n",
    "    \"\"\"\n",
    "    edges = seq_to_kmers(cycle, k=2)\n",
    "    values = []\n",
    "    for ea, eb in edges:\n",
    "        # v is heuristic “value” of edge\n",
    "        v = f(G, ea) + f(G, eb)\n",
    "        # Add value if cutting edge would isolate ea\n",
    "        if len(S(G, ea)) == 1:\n",
    "            v = v + f(G, ea)\n",
    "        # Add value if cutting edge would isolate eb\n",
    "        if len(P(G, eb)) == 1:\n",
    "            v = v + f(G, eb)\n",
    "        values.append(v)\n",
    "    ea, eb = edges[argmin(values)]\n",
    "    return ea, eb\n",
    "\n",
    "\n",
    "###############################################\n",
    "# Find optimal path through a graph of epitopes\n",
    "###############################################\n",
    "\n",
    "def find_optimal_path(G):\n",
    "    \"\"\"\n",
    "    Returns the optimal path through a graph of epitopes\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :returns: List of epitope strings on the optimal path\n",
    "    \"\"\"\n",
    "    # Forward loop - compute F(e)\n",
    "    for e in G.nodes:\n",
    "        F(G, e)\n",
    "    # Backward loop - build the path that achieves the maximal score\n",
    "    path = backward(G)\n",
    "    return path\n",
    "\n",
    "\n",
    "def F(G, e):\n",
    "    \"\"\"\n",
    "    Returns the maximum total score over all paths that end in e\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param e: String for a given potential T-cell epitope (PTE)\n",
    "    :returns: Float for the maximum total epitope score\n",
    "    \"\"\"\n",
    "    # Use precomputed F(e) if it already exists for the epitope\n",
    "    if 'F(e)' not in G.nodes[e]:\n",
    "        predecessors = P(G, e)\n",
    "        if not predecessors:\n",
    "            # If the set of predecessors P(e) is empty, then F(e) = f(e)\n",
    "            Fe = f(G, e)\n",
    "        else:\n",
    "            # If the set of predecessors P(e) is not empty, then F(e) = f(e) + max(F(P(e)))\n",
    "            Fe = f(G, e) + max([F(G, pe) for pe in predecessors])\n",
    "        # Save F(e) to the graph for this epitope\n",
    "        nx.set_node_attributes(G, {e: Fe}, 'F(e)')\n",
    "    return f(G, e, f='F(e)')\n",
    "\n",
    "\n",
    "def backward(G, path=[]):\n",
    "    \"\"\"\n",
    "    Returns the path that achieves the maximal score\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param path: List of epitope strings to complete (deafult=[])\n",
    "    :returns: List of epitope strings on path that achieve maximum score\n",
    "    \"\"\"\n",
    "    # Get the precomputed F(e) from the graph for all epitopes\n",
    "    Fe_dict = nx.get_node_attributes(G, 'F(e)')\n",
    "    if not path:\n",
    "        # Get the epitope with the maximum F(e) as the final epitope in our optimal path\n",
    "        end_nodes = {e: Fe_dict[e] for e in P(G, 'END')}\n",
    "        path = [max(end_nodes, key=end_nodes.get)]\n",
    "    # Get the most recently added epitope e and it's predecessors P(e)\n",
    "    e = path[0]\n",
    "    predecessors = P(G, e)\n",
    "    if predecessors[0] != 'BEGIN':\n",
    "        # Add the best (highest F(e)) predecessor P(e) of epitope e to our path\n",
    "        i = argmax([Fe_dict[pe] for pe in predecessors])\n",
    "        path.insert(0, predecessors[i])\n",
    "        # Repeat until you get to the start\n",
    "        backward(G, path)\n",
    "    return path\n",
    "\n",
    "\n",
    "###########################################################\n",
    "# Cocktail: Find (and iteratively refine) a set of antigens\n",
    "###########################################################\n",
    "\n",
    "def cocktail(G, m, refine=True, score='Score'):\n",
    "    \"\"\"\n",
    "    Returns a list of m antigens\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param m: Integer for number of antigens\n",
    "    :param refine: Boolean for if the antigens should be iteratively refined\n",
    "    :returns: List containing m antigens\n",
    "    \"\"\"\n",
    "    Q = [] # vaccine\n",
    "    # Save original epitope score so it can be reset later\n",
    "    score_dict = nx.get_node_attributes(G, score)\n",
    "    for n in range(0, m):\n",
    "        # Compute and save next antigen sequence\n",
    "        q = find_optimal_path(G)\n",
    "        # Add q to vaccine\n",
    "        Q.append(q)\n",
    "        # No credit for including e in subsequent antigens\n",
    "        for e in q:\n",
    "            nx.set_node_attributes(G, {e: 0}, score)\n",
    "        # Remove F(e) so it's recomputed using the updated scores\n",
    "        for (e,d) in G.nodes(data=True):\n",
    "            del d['F(e)']\n",
    "    # Reset to the original scores\n",
    "    nx.set_node_attributes(G, score_dict, score)\n",
    "    # Optional: Repeat - iterative refinement\n",
    "    if refine:\n",
    "        Q = iterative_refinement(G, Q)     \n",
    "    return Q\n",
    "\n",
    "\n",
    "def iterative_refinement(G, Q, score='Score'):\n",
    "    \"\"\"\n",
    "    Returns a list of iteratively refined antigens\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param Q: List containing antigens\n",
    "    :returns: List containing iteratively refined antigens\n",
    "    \"\"\"\n",
    "    m = len(Q)\n",
    "    # Save original epitope score so it can be reset later\n",
    "    score_dict = nx.get_node_attributes(G, score)\n",
    "    while True:\n",
    "        for n in range(0, m):\n",
    "            prev_Q = Q\n",
    "            # Remove sequence q from vaccine Q\n",
    "            q = Q[n]\n",
    "            Q.remove(q)\n",
    "            # No credit for including e in existing antigens\n",
    "            for q in Q:\n",
    "                for e in q:\n",
    "                    nx.set_node_attributes(G, {e: 0}, score)\n",
    "            # Compute replacement for old sequence q\n",
    "            q = find_optimal_path(G)\n",
    "            # Add q to vaccine\n",
    "            Q.insert(n, q)\n",
    "            # Reset to the original scores\n",
    "            nx.set_node_attributes(G, score_dict, score)\n",
    "            # Remove F(e) so it's recomputed using the updated scores\n",
    "            for (e,d) in G.nodes(data=True):\n",
    "                del d['F(e)']\n",
    "            if Q == prev_Q:\n",
    "                return Q\n",
    "\n",
    "\n",
    "##################################################\n",
    "# Construct/visualise the epitope graph and scores\n",
    "##################################################\n",
    "\n",
    "def construct_graph(kmers_dict, decycle=True, aligned=False, edge_colour='#BFBFBF'):\n",
    "    \"\"\"\n",
    "    Return a Directed Graph with unique k-mers as nodes, where overlapping k-mers are connected by edges\n",
    "    :param kmers_dict: Dictionary containing k-mers and their counts\n",
    "    :param decycle: Boolean if the output graph should be decycled (default=True)\n",
    "    :param edge_colour: String for edge colour (default='black')\n",
    "    :returns: Directed Graph containing epitopes\n",
    "    \"\"\"\n",
    "    # Create graph\n",
    "    G = nx.DiGraph()\n",
    "    # Add nodes - for each unique k-mer\n",
    "    for e, d in kmers_dict.items():\n",
    "        freq, immunogenicity, mhc_binding, evo_rate, pos, score = d['freq'], d['immunogenicity'], d['mhc_binding'], d['evo_rate'], d['pos'], d['score']\n",
    "        G.add_node(e, Frequency=freq, Immunogenicity=immunogenicity, MHC_Binding=mhc_binding, Evo_Rate=evo_rate, pos=(pos, score), Score=score)\n",
    "    # Add edges - where the last k−1 characters of ea match the first k−1 characters of eb\n",
    "    for ea, eb in product(G.nodes(), G.nodes()):\n",
    "        if not G.has_edge(ea, eb) and ea[1:] == eb[:-1]:\n",
    "            G.add_edge(ea, eb, colour=edge_colour)\n",
    "    # Decycle graph\n",
    "    if decycle:\n",
    "        G = decycle_graph(G)\n",
    "    # Add begin and end nodes\n",
    "    begin_nodes = [e for e in list(G.nodes) if not P(G, e)]\n",
    "    end_nodes = [e for e in list(G.nodes) if not S(G, e)]\n",
    "    G.add_node('BEGIN', Frequency=0, Immunogenicity=0, MHC_Binding=0, Evo_Rate=0, pos=(0,0), Score=0)\n",
    "    G.add_node('END', Frequency=0, Immunogenicity=0, MHC_Binding=0, Evo_Rate=0, pos=(0,0), Score=0)\n",
    "    for e in begin_nodes:\n",
    "        G.add_edge('BEGIN', e, colour=edge_colour)\n",
    "    for e in end_nodes:\n",
    "        G.add_edge(e, 'END', colour=edge_colour)\n",
    "    # Add the position node attribute (for unaligned sequences)\n",
    "    if not aligned:\n",
    "        for e in G.nodes():\n",
    "            if e != 'BEGIN' and e != 'END':\n",
    "                pos = nx.shortest_path_length(G, source='BEGIN', target=e)\n",
    "                score = f(G, e)\n",
    "                nx.set_node_attributes(G, {e: (pos, score)}, 'pos')\n",
    "    # Ensure the 'END' node is always at the end of the graph\n",
    "    end_pos = max([pos[0] for pos in list(nx.get_node_attributes(G, 'pos').values())]) + 1\n",
    "    nx.set_node_attributes(G, {'END': (end_pos, 0)}, 'pos')\n",
    "    return G\n",
    "\n",
    "\n",
    "def plot_graph(G, paths=None, node_size=150, with_labels=False, ylim=([0,1]), interactive=False, colour_by='Score'):\n",
    "    \"\"\"\n",
    "    Plot the epitope graph\n",
    "    :param G: Directed Graph containing epitopes\n",
    "    :param paths: List of lists of epitope strings (deafult=None)\n",
    "    :param node_size: Integer for size of nodes in non-interactive plot (default=150)\n",
    "    :param with_labels: Boolean for if epitope labels should be displayed in the non-interactive plot (default=False)\n",
    "    :param ylim: List for y-axis limits (default=[0,1])\n",
    "    :param interactive: Boolean for if the plot should be interactive (default=False)\n",
    "    :returns: None\n",
    "    \"\"\"\n",
    "    if interactive:\n",
    "        fig = ig.plot(G, color_method='Score', node_text=['Score', 'Frequency', 'Immunogenicity', 'MHC_Binding', 'pos']) # layout='spectral','spiral','spring'\n",
    "        return fig.show()\n",
    "    else:\n",
    "        # Define vars\n",
    "        node_color = list(nx.get_node_attributes(G,colour_by).values())\n",
    "        pos = nx.get_node_attributes(G, 'pos')\n",
    "        if paths:\n",
    "            for path in paths:\n",
    "                path = ['BEGIN'] + path + ['END']\n",
    "                for i in range(0,len(path)-1):\n",
    "                    G.edges[path[i], path[i+1]]['colour'] = 'red'\n",
    "        edge_colours = nx.get_edge_attributes(G, 'colour')\n",
    "        # Plot\n",
    "        fig, ax = plt.subplots(1,figsize=(16,8))\n",
    "        nx.draw(G, node_color=node_color, pos=pos, node_size=node_size, with_labels=with_labels, edge_color=edge_colours.values(), width=2, font_color='white', ax=ax)\n",
    "        limits=plt.axis('on')\n",
    "        max_pos = max([p[0] for p in pos.values()]) + 0.5\n",
    "        ax.set_xlim([-0.5, max_pos])\n",
    "        ax.set_ylim(ylim)\n",
    "        ax.spines.right.set_visible(False)\n",
    "        ax.spines.top.set_visible(False)\n",
    "        ax.tick_params(left=True, bottom=True, labelleft=True, labelbottom=True)\n",
    "        ax.tick_params(axis='both', which='major', labelsize=14)\n",
    "        plt.ylabel('Epitope Score, f(e)', fontsize=18)\n",
    "\n",
    "\n",
    "def plot_score(G, score='Frequency'):\n",
    "    \"\"\"\n",
    "    Plot the distribution of a given score for the nodes in the graph.\n",
    "    :param G: NetworkX graph\n",
    "    :param score: String for the score to be plotted (default='Frequency')\n",
    "    :returns: Figure\n",
    "    \"\"\"\n",
    "    scores = list(nx.get_node_attributes(G,score).values())\n",
    "    # Plot\n",
    "    fig, ax = plt.subplots(1,figsize=(16,8))\n",
    "    ax.hist(scores, bins=100)\n",
    "    ax.set_xlabel(score, fontsize=18)\n",
    "    ax.set_ylabel('Number of PTEs', fontsize=18)\n",
    "    ax.tick_params(axis='both', which='major', labelsize=14)\n",
    "    ax.spines.right.set_visible(False)\n",
    "    ax.spines.top.set_visible(False)\n",
    "    ax.tick_params(left=True, bottom=True, labelleft=True, labelbottom=True)\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple example "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the input epitopes\n",
    "# Cyclic example: kmers_dict = {'MSA': 0.6, 'SAM': 0.2, 'AMS': 0.4}\n",
    "kmers_dict = {\n",
    "    'MSA': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':1, 'freq':0, 'score':0.6},\n",
    "    'SAM': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':2, 'freq':0, 'score':0.2},\n",
    "    'AMQ': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':3, 'freq':0, 'score':0.2},\n",
    "    'MQL': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':4, 'freq':0, 'score':0.2},\n",
    "    'SAR': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':2, 'freq':0, 'score':0.4},\n",
    "    'MGA': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':1, 'freq':0, 'score':0.3},\n",
    "    'GAR': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':2, 'freq':0, 'score':0.7},\n",
    "    'ARQ': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':3, 'freq':0, 'score':0.4},\n",
    "    'RQL': {'evo_rate':0, 'immunogenicity':0, 'mhc_binding':0, 'pos':4, 'freq':0, 'score':0.4},\n",
    "}\n",
    "\n",
    "# Construct the graph\n",
    "G = construct_graph(kmers_dict, aligned=True)\n",
    "\n",
    "# Find the optimal path(s) through the graph of epitopes\n",
    "Q = cocktail(G, m)\n",
    "print([path_to_seq(path) for path in Q])\n",
    "\n",
    "# Plot the results\n",
    "plot_graph(G, paths=Q, node_size=2000, ylim=[-0.1, 1], with_labels=True, interactive=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform a Multiple Sequence Alignment (MSA)\n",
    "The input nucleotide sequences are preprocessed by performing a MSA. This is required to calculate omega (dN/dS) and is done in the following steps:\n",
    "1. **Pre-MSA:** Run these sequences through `pre-msa.bf` in order to correct frame-shift mutations and translate the resulting sequences to proteins\n",
    "2. **MSA:** Take the output of step 1 and run in through MAFFT to generate a **protein** MSA\n",
    "3. **Post-MSA:** Run the protein MSA (from step 2) and the frameshift corrected nucleotide sequences (from step 1) through `post-msa.bf` to obtain a nucleotide MSA. This step will also, compress all identical sequences, i.e. replace them with a single representative sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1) Pre-MSA\n",
    "# !hyphy {pre_msa_path} --input {fasta_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 2) MSA\n",
    "# !mafft --auto {fasta_pr_path} > {msa_pr_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 3) Post-MSA\n",
    "# !hyphy {post_msa_path} --protein-msa {msa_pr_path} --nucleotide-sequences {fasta_nt_path} --output {msa_nt_path}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare nonsynoymous (dN) and synonymous (dS) substitution rates\n",
    "\n",
    "Use HyPhy FEL to infer the number of nonsynoymous (N) and synonymous (S) substitutions on a per-site basis for a given coding alignment and corresponding phylogeny. Report N and S, and combine them to ascertain if N is greater than S:\n",
    "1. **Tree consutrction:** Consutruct a phylogenetic tree in NEWICK format using IQ-TREE\n",
    "2. **Compare dN and dS:** using the nucleotide MSA (from MSA step 3) and the phylogenetic tree (from step 1) as input to HyPhy FEL infer the nonsynoymous (dN), synonymous (dS) and estimate the number of synonymous/nonsynonymous sites for each codon/site in the MSA\n",
    "3. **Get Evolutionary rate:** load the JSON output from HyPhy SLAC (step 2) and adjust the sign (+/-) of the number of synonymous (-1) vs nonsynonymous (1) sites to indicate if the codon is under posotive/negative selection\n",
    "\n",
    "Note:\n",
    "- Diversifying or positive selection is where dS/α < dN/β\n",
    "- Purifying or negative selection is where dS/α > dN/β\n",
    "\n",
    "We want to select for codons under purifying selection because they're more conserved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1) Tree construction\n",
    "# !iqtree -s {msa_nt_path} -pre {fasta_path} --redo --fast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 2) Compare dN and dS\n",
    "# !hyphy slac --alignment {msa_nt_path} --tree {tree_path} --output {slac_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 3) Get Evolutionary Rate\n",
    "evo_rate = get_evo_rate(slac_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the FASTA sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if aligned:\n",
    "    fasta_path = msa_pr_path\n",
    "else:\n",
    "    fasta_path = fasta_pr_path\n",
    "\n",
    "fasta_seqs = SeqIO.parse(open(fasta_path),'fasta')\n",
    "seqs_dict = {seq.id:str(seq.seq) for seq in fasta_seqs}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split into k-mers\n",
    "Compute all possible k-mers of length `k` for the given target sequences and for each k-mer record the mean evolutionary rate, position in the alignment, epitope frequency and overall score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Add logic to use the input weights to check if all required inputs are present and only calculate desired scores\n",
    "if aligned:\n",
    "    df = seqs_dict_to_df(seqs_dict, evo_rate)\n",
    "    kmers_dict = seqs_df_to_kmers_dict(df, weights, k)\n",
    "else:\n",
    "    kmers_dict = seqs_to_kmers_dict(list(seqs_dict.values()), k=k, weights=weights, alleles=alleles, aligned=aligned)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct the epitope graph\n",
    "Create a Directed Graph (`DiGraph`) using the `networkx` package, where each epitope `e` is a node and edges connect nodes where the last `k−1` characters of `ea` match the first `k−1` characters of `eb`. For computational convenience, two extra nodes `BEGIN` and `END` are added. The `BEGIN` node connects to all the nodes that lack predecessors (`P(e)`) (corresponding to epitopes that are the first `k` characters in a sequence). Nodes that lack successors (`S(e)`) (because they are the last `k` characters in a sequence) are connected to the `END`. For plotting convenience, the length shortest path to the `BEGIN` node is added as a node attribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "G = construct_graph(kmers_dict, decycle=decycle, aligned=aligned)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assembly \n",
    "Take a path through the graph to optimise epitope frequency.\n",
    "\n",
    "The forward loop computes the function `F(e)` (the largest sum achievable for any path that terminates with the epitope `e`) for all the nodes in a stepwise manner. The backward loop chooses the node with maximum value as the final epitope in our optimal string and works backwards to build the path that achieves the maximal score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Add additional measures to scoring function eg binding affinity prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the optimal path(s) through the graph of epitopes\n",
    "Q = cocktail(G, m)\n",
    "[path_to_seq(path) for path in Q]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the epitope graph\n",
    "The nodes are the epitopes `e` and the edges connect epitopes whose sequences overlap by `k − 1` amino acids. The x-axis shows the shortest path length to the `BEGIN` node, the y-axis indicates the epitope frequency `f(e)` in this target sequence set. The optimal path is shown in red which corresponds to the protein sequence that maximizes epitope coverage of the population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# F(e)\n",
    "plot_graph(G, paths=Q, node_size=25, with_labels=False, ylim=[0, 1], interactive=False, colour_by='Score')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluate the implemented scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the scores of each node in the path\n",
    "pos = [G.nodes[node]['pos'][0] for node in Q[0]]\n",
    "freq = [G.nodes[node]['Frequency'] for node in Q[0]]\n",
    "mhc_binding = [G.nodes[node]['MHC_Binding'] for node in Q[0]]\n",
    "immunogenicity = [G.nodes[node]['Immunogenicity'] for node in Q[0]]\n",
    "\n",
    "# Plot stacked area\n",
    "fig, ax = plt.subplots(1,figsize=(16,8))\n",
    "ax.stackplot(pos, freq, mhc_binding, immunogenicity, labels=['Frequency', 'MHC Binding', 'Immunogenicity'])\n",
    "ax.set_xlabel('Position', fontsize=18)\n",
    "ax.set_ylabel('Score', fontsize=18)\n",
    "ax.tick_params(axis='both', which='major', labelsize=14)\n",
    "ax.spines.right.set_visible(False)\n",
    "ax.spines.top.set_visible(False)\n",
    "ax.tick_params(left=True, bottom=True, labelleft=True, labelbottom=True)\n",
    "ax.legend(loc='upper left', fontsize=14)\n",
    "ax.set_xlim([0, max(pos)])\n",
    "plt.show()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "View the distribution of the scores for all the epitopes in the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist_plot = plot_score(G, score='Frequency')\n",
    "# hist_plot = plot_score(G, score='MHC_Binding')\n",
    "# hist_plot = plot_score(G, score='Immunogenicity')\n",
    "hist_plot.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare the values for different scores to see if each score is providing additional information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "freq_dict = nx.get_node_attributes(G, 'Frequency')\n",
    "freqs = list(freq_dict.values())\n",
    "\n",
    "immunogenicity_dict = nx.get_node_attributes(G, 'Immunogenicity')\n",
    "immunogenecities = list(immunogenicity_dict.values())\n",
    "\n",
    "mhc_binding_dict = nx.get_node_attributes(G, 'MHC_Binding')\n",
    "mhc_bindings = list(mhc_binding_dict.values())\n",
    "\n",
    "plt.scatter(x=immunogenecities,y=mhc_bindings)\n",
    "plt.xlabel('immunogenecities')\n",
    "plt.ylabel('MHC_Binding')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tvax",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "f873ee3bbd6b65e3060a1ff5e0b4755f776b262fd9907a4f86f67620d9e561eb"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
